v2/inc/c_logging/log_context.h

```diff
@@ -4,6 +4,241 @@
+    /* Codes_SRS_LOG_CONTEXT_01_005: [ LOG_CONTEXT_PROPERTY shall expand to code copying the value property_value to be the value of the property/value pair. ]*/ \
+    (void)property_type##_log_context_property_type_init((void*)data_pos, field_value); \
+    data_pos += sizeof(property_type); \
+    property_value_pair++; \
+
+#define SETUP_PROPERTY_PAIR(field_desc) \
+    MU_C2(EXPAND_SETUP_PROPERTY_PAIR_, field_desc)
+
+// COUNT_PROPERTY
+
+#define EXPAND_COUNT_PROPERTY_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+    + 1
+
+#define EXPAND_COUNT_PROPERTY_LOG_CONTEXT_NAME(log_context_name) \
+
+#define EXPAND_COUNT_PROPERTY_LOG_CONTEXT_PROPERTY(property_type, property_name, field_value) \
+    + 1
+
+#define COUNT_PROPERTY(field_desc) \
+    MU_C2(EXPAND_COUNT_PROPERTY_, field_desc)
+
+// COUNT_DATA_BYTES
+
+#define EXPAND_COUNT_DATA_BYTES_LOG_CONTEXT_STRING_PROPERTY(property_name, ...) \
+    + snprintf(NULL, 0, __VA_ARGS__) + 1
+
+#define EXPAND_COUNT_DATA_BYTES_LOG_CONTEXT_NAME(log_context_name) \
+
+#define EXPAND_COUNT_DATA_BYTES_LOG_CONTEXT_PROPERTY(property_type, property_name, field_value) \
+    + sizeof(property_type)
+
+#define COUNT_DATA_BYTES(field_desc) \
+    MU_C2(EXPAND_COUNT_DATA_BYTES_, field_desc)
+
+// Macro that can be used to create a context on the stack
+// We allocate on the stack enough space for a max payload for the context.
+#define LOG_CONTEXT_LOCAL_DEFINE(destination_context, parent_context, ...) \
+    uint8_t MU_C2(values_log_data_, destination_context)[LOG_MAX_STACK_DATA_SIZE]; \
+    LOG_CONTEXT destination_context; \
+    MU_IF(MU_COUNT_ARG(__VA_ARGS__), LOG_CONTEXT_CHECK_VARIABLE_ARGS(__VA_ARGS__),) \
+    LOG_CONTEXT_PROPERTY_VALUE_PAIR MU_C2(property_values_pair_, destination_context)[LOG_MAX_STACK_PROPERTY_VALUE_PAIR_COUNT]; \
+    { \
+        destination_context.values_data = MU_C2(values_log_data_, destination_context); \
+        destination_context.values_data_length = internal_log_context_get_values_data_length_or_zero(parent_context) + 1 MU_IF(MU_COUNT_ARG(__VA_ARGS__), MU_FOR_EACH_1(COUNT_DATA_BYTES, __VA_ARGS__),); \
+        destination_context.property_value_pairs_ptr = MU_C2(property_values_pair_, destination_context); \
+        destination_context.property_value_pair_count = internal_log_context_get_property_value_pair_count_or_zero(parent_context) + 1 MU_IF(MU_COUNT_ARG(__VA_ARGS__), MU_FOR_EACH_1(COUNT_PROPERTY, __VA_ARGS__),); \
+        /* Codes_SRS_LOG_CONTEXT_01_024: [ If the number of properties to be stored in the log context exceeds LOG_MAX_STACK_PROPERTY_VALUE_PAIR_COUNT, an error shall be reported by calling log_internal_error_report and no properties shall be stored in the context. ]*/ \
+        if (destination_context.property_value_pair_count > LOG_MAX_STACK_PROPERTY_VALUE_PAIR_COUNT) \
+        { \
+            (void)printf("Too many properties: property_value_pair_count = %" PRIu32 "\r\n", destination_context.property_value_pair_count); \
```

Comment:
>oo many properties [](http://example.com/codeflow?start=27&length=18)

If we trim string outputs, maybe we can trim the number of properties and instead of goign belly up, just continue with the "whatever we could take"? (and still call log_internal_error_report).

Reply:
Yeah, I thought about that one for a bit too and maybe it is a bit more involved. And that's when I said probably should do that as a separate task and go with belly up for now as this PR was reaching some proportions I do not like!
Task is here in the current PBI: 
ID Work Item Type Title State Assigned To Effort Remaining Work Original Estimate Value Area Iteration Path Tags
15453524 Task When the max stack properties for a log context is exceeded we should truncate the properties rather than go belly up To Do           One\Custom\Azure Messaging Store\Iteration 271